server/apps/immich/src/api-v1/share/share.core.ts

```diff
@@ -0,0 +1,76 @@
+import { SharedLinkEntity } from '@app/database/entities/shared-link.entity';
+import { CreateSharedLinkDto } from './dto/create-shared-link.dto';
+import { ISharedLinkRepository } from './shared-link.repository';
+import crypto from 'node:crypto';
+import { BadRequestException, InternalServerErrorException, Logger } from '@nestjs/common';
+import { AssetEntity } from '@app/database';
+
+export class ShareCore {
+  readonly logger = new Logger(ShareCore.name);
+
+  constructor(private sharedLinkRepository: ISharedLinkRepository) {}
+
+  async createSharedLink(userId: string, dto: CreateSharedLinkDto): Promise<SharedLinkEntity> {
+    try {
+      const sharedLink = new SharedLinkEntity();
+
+      sharedLink.id = Buffer.from(crypto.randomBytes(70));
+      sharedLink.key = Buffer.from(crypto.randomBytes(50));
+      sharedLink.description = dto.description;
+      sharedLink.userId = userId;
+      sharedLink.createdAt = new Date().toISOString();
+      sharedLink.expiresAt = dto.expiredAt;
+      sharedLink.type = dto.sharedType;
+      sharedLink.assets = dto.assets;
+      sharedLink.album = dto.album;
+      sharedLink.allowUpload = dto.allowUpload ?? false;
+
+      return this.sharedLinkRepository.create(sharedLink);
+    } catch (error: any) {
+      this.logger.error(error, error.stack);
+      throw new InternalServerErrorException('failed to create shared link');
+    }
+  }
+
+  async getSharedLinks(userId: string): Promise<SharedLinkEntity[]> {
+    return this.sharedLinkRepository.get(userId);
+  }
+
+  async removeSharedLink(id: string): Promise<SharedLinkEntity> {
+    const link = await this.getSharedLinkById(id);
+
+    if (!link) {
+      throw new BadRequestException('Shared link not found');
+    }
+
+    return await this.sharedLinkRepository.remove(link);
+  }
+
+  async getSharedLinkById(id: string): Promise<SharedLinkEntity> {
+    const link = await this.sharedLinkRepository.getbyId(id);
+
+    if (!link) {
+      throw new BadRequestException('Shared link not found');
+    }
+
+    return link;
+  }
+
+  async getSharedLinkByKey(key: string): Promise<SharedLinkEntity> {
+    const link = await this.sharedLinkRepository.getByKey(key);
+
+    if (!link) {
+      throw new BadRequestException();
+    }
+
+    return link;
+  }
+
+  async updataAssetsInSharedLink(sharedLinkId: string, assets: AssetEntity[]) {
+    const link = await this.getSharedLinkById(sharedLinkId);
+
+    link.assets = assets;
+
+    return await this.sharedLinkRepository.save(link);
+  }
```

Comment:
Does this need authentication or a check the user has access to this link?

Reply:
It will be, I haven't made an endpoint to edit the share link yet.